**这是用来读取spei nc文件的测试代码**

In [42]:
# === 导入必要的库 ===
import xarray as xr
import geopandas as gpd
import pandas as pd
import numpy as np
from shapely.geometry import Point
import matplotlib.pyplot as plt
import seaborn as sns

### 加载 SPEI 数据
我们将加载 SPEI 的 NetCDF 数据文件，并提取时间、经纬度和 SPEI 值。

In [43]:
# === 加载 SPEI 数据 ===
spei_nc_path = "../data/raw/spei03.nc"

# 加载 NetCDF 数据
ds = xr.open_dataset(spei_nc_path)
spei = ds['spei']
lats = ds['lat'].values
lons = ds['lon'].values
times = pd.to_datetime(ds['time'].values)

# 打印数据基本信息
print("纬度范围:", lats.min(), "-", lats.max())
print("经度范围:", lons.min(), "-", lons.max())
print("时间范围:", times.min(), "-", times.max())

纬度范围: -89.75 - 89.75
经度范围: -179.75 - 179.75
时间范围: 1901-01-16 00:00:00 - 2023-12-16 00:00:00


### 读取国家边界矢量文件
我们将加载国家边界的 shapefile 文件，并提取国家名称和 ISO3 代码。

In [45]:
# === 读取国家边界矢量文件 ===
shapefile_path = '../data/raw/ne_50m_admin_0_countries/ne_50m_admin_0_countries.shp'
countries = gpd.read_file(shapefile_path).to_crs('EPSG:4326')

# 提取国家名称和 ISO3 代码
namelist = countries[['ADMIN', 'ISO_A3']].drop_duplicates().reset_index(drop=True)
namelist = namelist.rename(columns={'ADMIN': 'country_name', 'ISO_A3': 'iso3'})

# 保存到 processed 文件夹
namelist.to_csv('../data/processed/ne_50m_admin_0_countries_namelist.csv', index=False)
print("国家名称和 ISO3 代码已保存到: ne_50m_admin_0_countries_namelist.csv")

DataSourceError: ../data/raw/ne_50m_admin_0_countries/ne_50m_admin_0_countries.shp: No such file or directory

ne_50m_admin_0_countries_namelist.csv里有242个国家，但是这个数据的质量有些问题，有些国家的iso3代码为-99

### 构建 SPEI 网格点 GeoDataFrame
将 SPEI 数据的经纬度网格点转换为 GeoDataFrame，以便进行空间连接。

In [ ]:
# === 构建 SPEI 网格点 GeoDataFrame ===
lon_grid, lat_grid = np.meshgrid(lons, lats)
points = [Point(lon, lat) for lon, lat in zip(lon_grid.flatten(), lat_grid.flatten())]
points_gdf = gpd.GeoDataFrame({'orig_idx': np.arange(len(points))}, geometry=points, crs='EPSG:4326')

### 空间连接
将每个网格点分配到对应的国家。

In [ ]:
# === 空间连接 ===
points_gdf = gpd.sjoin(points_gdf, countries[['ADMIN', 'ISO_A3', 'geometry']], how='inner', predicate='within')
points_gdf = points_gdf.rename(columns={'ADMIN': 'country', 'ISO_A3': 'iso3'})
print("空间连接完成，示例数据：")
points_gdf.head()

空间连接完成，示例数据：


,orig_idx,geometry,index_right,country,iso3
0,0,POINT (-179.75 -89.75),239,Antarctica,ATA
1,1,POINT (-179.25 -89.75),239,Antarctica,ATA
2,2,POINT (-178.75 -89.75),239,Antarctica,ATA
3,3,POINT (-178.25 -89.75),239,Antarctica,ATA
4,4,POINT (-177.75 -89.75),239,Antarctica,ATA


### 提取每国每月 SPEI 均值
计算 2019-2022 年每个国家每月的 SPEI 平均值。

In [ ]:
# === 提取每国每月 SPEI 均值 ===
records = []
for t_idx, t in enumerate(times):
    if t.year < 2019 or t.year > 2022:
        continue
    spei_slice = spei.isel(time=t_idx).values.flatten()
    points_gdf['spei'] = spei_slice[points_gdf['orig_idx'].values]
    grouped = points_gdf.groupby(['country', 'iso3'])['spei'].mean().reset_index()
    grouped['spei_date'] = t
    records.append(grouped)

result = pd.concat(records, ignore_index=True)
result = result.dropna(subset=['country'])

In [ ]:
result
result.to_csv('../data/processed/spei_result_preview.csv', index=False)
print("已保存 result 的预览到 result_preview.csv")

已保存 result 的预览到 result_preview.csv


### 数据检查
检查提取的 SPEI 数据是否完整，并修正可能的错误。

In [ ]:
# 1. shp文件所有国家/地区
all_admin = set(countries['ADMIN'].unique())
print(f"shp文件中ADMIN唯一值数量: {len(all_admin)}")

# 2. 空间连接后country的唯一值
joined_admin = set(result['country'].unique())
print(f"空间连接后country唯一值数量: {len(joined_admin)}")

# 3. 哪些shp里的国家没有出现在result里
missing_admin = all_admin - joined_admin
print(f"shp中但空间连接后缺失的国家/地区: {list(missing_admin)}")

# 4. result中ISOO3为-99或缺失的情况
print("result中ISOO3为-99的国家：", result.loc[result['iso3'] == '-99', 'country'].unique())
print("result中ISOO3缺失的数量：", result['iso3'].isnull().sum())

# 5. 检查result中country和iso3的对应关系
print("country和iso3一对多的情况：")
print(result.groupby('iso3')['country'].nunique().value_counts())

shp文件中ADMIN唯一值数量: 242
空间连接后country唯一值数量: 190
shp中但空间连接后缺失的国家/地区: ['Turks and Caicos Islands', 'Hong Kong S.A.R.', 'Tuvalu', 'American Samoa', 'Bahrain', 'Vatican', 'Macao S.A.R', 'Jersey', 'Indian Ocean Territories', 'Saint Pierre and Miquelon', 'Guernsey', 'Andorra', 'Montserrat', 'Maldives', 'Palau', 'Curaçao', 'Dominica', 'Sint Maarten', 'British Indian Ocean Territory', 'Singapore', 'Norfolk Island', 'Saint Lucia', 'Seychelles', 'São Tomé and Principe', 'Niue', 'San Marino', 'Cook Islands', 'Ashmore and Cartier Islands', 'Grenada', 'Barbados', 'Nauru', 'Tonga', 'Anguilla', 'Antigua and Barbuda', 'Pitcairn Islands', 'Northern Mariana Islands', 'Bermuda', 'Saint Barthelemy', 'Saint Kitts and Nevis', 'British Virgin Islands', 'Monaco', 'Isle of Man', 'Malta', 'Liechtenstein', 'Federated States of Micronesia', 'Wallis and Futuna', 'Guam', 'Saint Helena', 'Heard Island and McDonald Islands', 'Aruba', 'Saint Martin', 'Marshall Islands']
result中ISOO3为-99的国家： ['France' 'Kosovo' 'Northern C

In [ ]:
# 2. 检查空间连接后缺失的国家的多边形面积
missing_countries_gdf = countries[countries['ADMIN'].isin(missing_admin)].copy()
missing_countries_gdf['area'] = missing_countries_gdf['geometry'].to_crs('EPSG:3857').area / 1e6  # 单位：平方公里
print(missing_countries_gdf[['ADMIN', 'area']].sort_values('area'))

# 3. 检查是否有SPEI网格点落入这些国家（带序号输出）
for idx, admin in enumerate(missing_admin, 1):
    poly = missing_countries_gdf[missing_countries_gdf['ADMIN'] == admin].geometry.iloc[0]
    n_points = points_gdf.within(poly).sum()
    print(f"{idx}. {admin} 内的SPEI网格点数: {n_points}")

# 获取缺失国家的名字和ISO_A3
missing_countries_gdf = countries[countries['ADMIN'].isin(missing_admin)][['ADMIN', 'ISO_A3']]
missing_countries_gdf = missing_countries_gdf.reset_index(drop=True)
print(missing_countries_gdf)
missing_countries_gdf.to_csv('../data/processed/spei_missing_countries.csv', index=False)


                                 ADMIN         area
5                              Vatican     1.278113
229        Ashmore and Cartier Islands     2.856467
241                             Tuvalu    10.503433
108                             Monaco    23.847567
196                        Macao S.A.R    26.595111
164                   Saint Barthelemy    27.142060
100                              Nauru    27.859625
240                       Sint Maarten    46.381684
20                    Pitcairn Islands    46.713122
228                     Norfolk Island    53.958729
163                       Saint Martin    54.358330
114                           Maldives    66.847959
95                        Cook Islands    69.953413
27                          Montserrat    87.193406
21                            Anguilla    92.520471
24                             Bermuda    92.584186
29                            Guernsey   114.037389
14                      American Samoa   128.520951
69          

下面代码解决country和ISO_A3一对多的情况，也就是：result中ISO_A3为-99的国家： ['France' 'Kosovo' 'Northern Cyprus' 'Norway' 'Siachen Glacier'
 'Somaliland']，其中Northern Cyprus不被联合国承认，所以没有iso-A3，我们这里直接删除；其他的手动匹配

In [ ]:
# 直接从result中删除Northern Cyprus
result = result[result['country'] != 'Northern Cyprus'].copy()

In [ ]:
# 1. 构建手动映射表（示例，需根据实际情况补充）
iso3_manual_map = {
    'France': 'FRA',
    'Kosovo': 'XKX',
    'Norway': 'NOR',
    'Siachen Glacier': 'SIA',
    'Somaliland': 'SOL'
}

# 2. 替换spei_clean中的ISO_A3
def fix_iso3(row):
    if row['iso3'] == '-99':
        return iso3_manual_map.get(row['country'], None)
    else:
        return row['iso3']

result = result[result['country'] != 'Northern Cyprus'].copy()
result['iso3_fixed'] = result.apply(fix_iso3, axis=1)

# 3. 检查修正结果
print(result[result['iso3'] == '-99'][['country', 'iso3', 'iso3_fixed']])

# 4. 用修正后的iso3替换原有列
result['iso3'] = result['iso3_fixed']
result = result.drop(columns=['iso3_fixed'])


              country iso3 iso3_fixed
57             France  -99        FRA
89             Kosovo  -99        XKX
124            Norway  -99        NOR
146   Siachen Glacier  -99        SIA
152        Somaliland  -99        SOL
...               ...  ...        ...
8987           France  -99        FRA
9019           Kosovo  -99        XKX
9054           Norway  -99        NOR
9076  Siachen Glacier  -99        SIA
9082       Somaliland  -99        SOL

[240 rows x 3 columns]


In [ ]:
# 检查新的result，统计国家和iso3的唯一数量
print("去除Northern Cyprus后，result中country唯一值数量：", result['country'].nunique())
print("去除Northern Cyprus后，result中iso3唯一值数量：", result['iso3'].nunique())

去除Northern Cyprus后，result中country唯一值数量： 189
去除Northern Cyprus后，result中iso3唯一值数量： 189


检查完成：
1. 上面52个没有连接上的国家，是因为面积太小，没有spei网格点落入；
2. 修正了5个ISO-3错误的国家代码（-99->real ISO-A3），删除了北塞浦路斯这个不被联合国承认的国家
3. 242-52-1 = 189
3. 目前的result是都有至少一个网格点落入国界的，但是由于spei可能存在空值，所以需要再次检查。

In [ ]:
# 检查整体缺失
print("总行数：", len(result))
print("每列缺失值数量：\n", result.isnull().sum())

# 检查哪些国家有缺失
missing_country = result[result['spei'].isnull()]
print("有空值的国家数量：", missing_country['country'].nunique())
print("有空值的国家代码：", missing_country['country'].unique())

总行数： 9072
每列缺失值数量：
 country        0
iso3           0
spei         144
spei_date      0
dtype: int64
有空值的国家数量： 3
有空值的国家代码： ['Antarctica' 'Cayman Islands' 'Kiribati']


In [ ]:
# 检查Kiribati
country_poly = countries[countries['ADMIN'] == 'Kiribati'].geometry.iloc[0]
points_in_country = points_gdf[points_gdf.within(country_poly)]
print(f"Kiribati内的网格点数: {len(points_in_country)}")

# 找到该点的经纬度
pt = points_in_country.geometry.iloc[0]
lon, lat = pt.x, pt.y

# 找到最近的SPEI网格索引
lat_idx = np.argmin(np.abs(lats - lat))
lon_idx = np.argmin(np.abs(lons - lon))

# 检查所有时间的SPEI值
spei_values = spei[:, lat_idx, lon_idx].values
print("该点所有时间的SPEI值是否全为NaN：", np.all(np.isnan(spei_values)))
print("部分SPEI值：", spei_values[:10])

Kiribati内的网格点数: 1
该点所有时间的SPEI值是否全为NaN： True
部分SPEI值： [nan nan nan nan nan nan nan nan nan nan]


In [ ]:
# 检查Cayman Islands
country_poly_CI = countries[countries['ADMIN'] == 'Cayman Islands'].geometry.iloc[0]
points_in_country_CI = points_gdf[points_gdf.within(country_poly_CI)]
print(f"Cayman Islands内的网格点数: {len(points_in_country_CI)}")

# 找到该点的经纬度
pt_CI = points_in_country_CI.geometry.iloc[0]
lon_CI, lat_CI = pt_CI.x, pt_CI.y

# 找到最近的SPEI网格索引
lat_idx_CI = np.argmin(np.abs(lats - lat_CI))
lon_idx_CI = np.argmin(np.abs(lons - lon_CI))

# 检查所有时间的SPEI值
spei_values_CI = spei[:, lat_idx_CI, lon_idx_CI].values
print("该点所有时间的SPEI值是否全为NaN：", np.all(np.isnan(spei_values_CI)))
print("部分SPEI值：", spei_values_CI[:10])


Cayman Islands内的网格点数: 1
该点所有时间的SPEI值是否全为NaN： True
部分SPEI值： [nan nan nan nan nan nan nan nan nan nan]


In [ ]:
# 检查Antarctica
country_poly_A = countries[countries['ADMIN'] == 'Antarctica'].geometry.iloc[0]
points_in_country_A = points_gdf[points_gdf.within(country_poly_A)]
print(f"Antarctica内的网格点数: {len(points_in_country_A)}")

# 找到该点的经纬度
pt_A = points_in_country_A.geometry.iloc[0]
lon_A, lat_A = pt_A.x, pt_A.y

# 找到最近的SPEI网格索引
lat_idx_A = np.argmin(np.abs(lats - lat_A))
lon_idx_A = np.argmin(np.abs(lons - lon_A))

# 检查所有时间的SPEI值
spei_values_A = spei[:, lat_idx_A, lon_idx_A].values
print("该点所有时间的SPEI值是否全为NaN：", np.all(np.isnan(spei_values_A)))
print("部分SPEI值：", spei_values_A[:10])


Antarctica内的网格点数: 24165
该点所有时间的SPEI值是否全为NaN： True
部分SPEI值： [nan nan nan nan nan nan nan nan nan nan]


所以检查的结果是，三个地方的都有spei网格点，但是spei的值确实都是Nan

In [ ]:
# 删除有缺失的三个国家
drop_countries = ['Antarctica', 'Cayman Islands', 'Kiribati']
spei_clean = result[~result['country'].isin(drop_countries)].copy()

# 检查是否还有缺失
print("每列缺失值数量：\n", spei_clean.isnull().sum())
print("总行数：", len(spei_clean))

每列缺失值数量：
 country      0
iso3         0
spei         0
spei_date    0
dtype: int64
总行数： 8928


In [ ]:
spei_clean

,country,iso3,spei,spei_date
0,Afghanistan,AFG,-0.101823,2019-01-16
1,Aland,ALA,-0.584555,2019-01-16
2,Albania,ALB,-0.229678,2019-01-16
3,Algeria,DZA,-1.081419,2019-01-16
4,Angola,AGO,0.090749,2019-01-16
...,...,...,...,...
9115,Western Sahara,ESH,-1.458937,2022-12-16
9116,Yemen,YEM,-1.207492,2022-12-16
9117,Zambia,ZMB,0.084212,2022-12-16
9118,Zimbabwe,ZWE,-0.164138,2022-12-16


现在完成了数据的清理，得到的spei_country_month_2019_2022_cleaned.csv是没有缺失值的，ISO3正确且完整的

In [ ]:
# 最终数据检查

print("时间范围：", spei_clean['spei_date'].min(), "到", spei_clean['spei_date'].max())
print("spei列是否有空值：", spei_clean['spei'].isnull().sum() == 0)
print("iso3列是否有空值：", spei_clean['iso3'].isnull().sum() == 0)
print("iso3是否还有-99：", (spei_clean['iso3'] == '-99').sum() == 0)
print("国家数量：", spei_clean['country'].nunique())
print("iso3数量：", spei_clean['iso3'].nunique())

# 每个国家的spei数量
spei_count = spei_clean.groupby('country')['spei'].count()
print("每个国家的spei数量是否都是48：", (spei_count == 48).all())
print("spei数量不是48的国家：")
print(spei_count[spei_count != 48])

时间范围： 2019-01-16 00:00:00 到 2022-12-16 00:00:00
spei列是否有空值： True
iso3列是否有空值： True
iso3是否还有-99： True
国家数量： 186
iso3数量： 186
每个国家的spei数量是否都是48： True
spei数量不是48的国家：
Series([], Name: spei, dtype: int64)


In [ ]:
# === 修改并保存数据 ===

# 拆分 spei_date 为 year 和 month
spei_clean['year'] = spei_clean['spei_date'].dt.year
spei_clean['month'] = spei_clean['spei_date'].dt.month

# 去掉 country 列，重命名 iso3 为 spei_iso3
spei_clean = spei_clean.drop(columns=['country'])
spei_clean = spei_clean.rename(columns={'iso3': 'spei_iso3'})

# 保存修改后的数据
spei_clean.to_csv('../data/processed/spei_country_month.csv', index=False)
print("最终面板已保存，无空值，每国48期，ISO3改名为 spei_iso3，保存为 spei_country_month.csv")

最终面板已保存，无空值，每国48期，ISO3改名为 spei_iso3，保存为 spei_country_month.csv


以上是全部有效代码

### 对比py文件
下面的代码是对对应的py文件输出文件的检查，确保ipynb和py文件输出一致

In [ ]:
import pandas as pd

# 读取两个文件
df1 = pd.read_csv('../data/processed/spei03_country_month_cleaned.csv')
df2 = pd.read_csv('../data/processed/spei_country_month_test.csv')

def quick_info(df, name):
    print(f"==== {name} ====")
    print("shape:", df.shape)
    print("国家数：", df['country'].nunique())
    print("ISO3数：", df['ISO_A3'].nunique())
    print("时间范围：", df['date'].min(), "到", df['date'].max())
    print("spei有空值？", df['spei'].isnull().sum() > 0)
    print("ISO_A3有空值？", df['ISO_A3'].isnull().sum() > 0)
    print("ISO_A3有-99？", (df['ISO_A3'] == '-99').sum() > 0)
    print("每国spei数量都为48？", (df.groupby('country')['spei'].count() == 48).all())
    print()

quick_info(df1, "spei03_country_month_cleaned.csv")
quick_info(df2, "spei_country_month_test.csv")

# 可选：对比国家列表差异
print("df1独有国家：", set(df1['country']) - set(df2['country']))
print("df2独有国家：", set(df2['country']) - set(df1['country']))

==== spei03_country_month_cleaned.csv ====
shape: (8928, 4)
国家数： 186
ISO3数： 186
时间范围： 2019-01-16 到 2022-12-16
spei有空值？ False
ISO_A3有空值？ False
ISO_A3有-99？ False
每国spei数量都为48？ True

==== spei_country_month_test.csv ====
shape: (8928, 4)
国家数： 186
ISO3数： 186
时间范围： 2019-01-16 到 2022-12-16
spei有空值？ False
ISO_A3有空值？ False
ISO_A3有-99？ False
每国spei数量都为48？ True

df1独有国家： set()
df2独有国家： set()

shape: (8928, 4)
国家数： 186
ISO3数： 186
时间范围： 2019-01-16 到 2022-12-16
spei有空值？ False
ISO_A3有空值？ False
ISO_A3有-99？ False
每国spei数量都为48？ True

==== spei_country_month_test.csv ====
shape: (8928, 4)
国家数： 186
ISO3数： 186
时间范围： 2019-01-16 到 2022-12-16
spei有空值？ False
ISO_A3有空值？ False
ISO_A3有-99？ False
每国spei数量都为48？ True

df1独有国家： set()
df2独有国家： set()


输出一致，可以使用了